InnerEye/ML/baselines_util.py

```diff
@@ -173,3 +184,119 @@ def get_comparison_baselines(outputs_folder: Path, azure_config: AzureConfig,
+            return CONTENTS_MISMATCH
+    else:
+        expected_binary = expected.read_bytes()
+        actual_binary = actual.read_bytes()
+        if expected_binary != actual_binary:
+            logging.debug(f"Expected {len(expected_binary)} bytes, actual {len(actual_binary)} bytes")
+            return CONTENTS_MISMATCH
+    return ""
+
+
+def compare_folder_contents(expected_folder: Path,
+                            actual_folder: Optional[Path] = None,
+                            run: Optional[Run] = None) -> None:
+    """
+    Compares a set of files in a folder, against files in either the other folder or files stored in the given
+    AzureML run. Each file that is present in the "expected" folder must be also present in the "actual" folder
+    (or the AzureML run), with exactly the same contents, in the same folder structure.
+    For example, if there is a file "<expected>/foo/bar/contents.txt", then there must also be a file
+    "<actual>/foo/bar/contents.txt"
+    If a file is missing, or does not have the expected contents, an exception is raised.
+    :param expected_folder: A folder with files that are expected to be present.
+    :param actual_folder: The output folder with the actually produced files.
+    :param run: An AzureML run
+    """
+    logging.debug(f"Checking job output against expected files in folder {expected_folder}")
+    logging.debug(f"Current working directory: {Path.cwd()}")
+    messages = []
+    if not expected_folder.is_dir():
+        raise ValueError(f"Folder with expected files does not exist: {expected_folder}")
+    if run and is_offline_run_context(run):
+        logging.warning("Skipping file comparison because the given run context is an AzureML offline run.")
+        return
+    files_in_run: List[str] = run.get_file_names() if run else []
+    temp_folder = Path(tempfile.mkdtemp()) if run else None
+    for file in expected_folder.rglob("*"):
+        # rglob also returns folders, skip those
+        if file.is_dir():
+            continue
+        logging.debug(f"Checking file {file}")
+        # All files stored in AzureML runs use Linux-style path
+        file_relative = file.relative_to(expected_folder).as_posix()
+        if str(file_relative).startswith(REGRESSION_TEST_AZUREML_FOLDER) or \
+                str(file_relative).startswith(REGRESSION_TEST_AZUREML_PARENT_FOLDER):
+            continue
+        actual_file: Optional[Path] = None
+        if actual_folder:
+            actual_file = actual_folder / file_relative
+            if not actual_file.is_file():
+                actual_file = None
+        elif temp_folder is not None and run is not None:
+            if file_relative in files_in_run:
+                actual_file = temp_folder / file_relative
+                run.download_file(name=str(file_relative), output_file_path=str(actual_file))
+        message = compare_files(expected=file, actual=actual_file) if actual_file else "Missing file"
```

Comment:
This should be if actual_file.exists()

Reply:
In the case of downloading from runs, run.download ensures that the file exists. In the other case, we do the check in line 260, and set it to null if it does not exist. Agreed, it's a bit twisted